pyClient/zeth/joinsplit.py

```diff
@@ -265,6 +265,38 @@ def sign(
     return sigma
 
 
+def sign_joinsplit(
+        joinsplit_keypair: JoinsplitKeypair,
+        pk_sender: bytes,
+        ciphertexts: List[bytes],
+        proof_json: Dict[str, Any]) -> int:
+
```

Comment:
Please add a comment for documentation to say what the function does in a few words.
Also, why do you pass the entire keypair? You only need the signing key, not the verification key here

Reply:
Comment added.

We need the verification key for the signature (the challenge is a hash on the message and verification key, we do not sign the verification key however).

Reply:
Right, of course you're right, vk needs to be hashed. Why is the full key pair passed to the `sign` function here: `joinsplit_sig = sign(joinsplit_keypair, hash_to_be_sign)`?

Reply:
For the same reason.

sign creates a signature on a hash (so the challenge is computed in here) while the new function creates a signature on the hash of the primary inputs, ciphertexts, pk_sender and proof.

Reply:
That's misleading. I'll check the code of the function, but by reading at the code you expect `sign` to be a signature algorithm `Sign(signing_key, message_to_sign)`.

Reply:
The `sign` function goes like:
```
def sign(
        keypair: JoinsplitKeypair,
        hash_signature: str) -> int:
    """
    Generate a Schnorr one-time signature of the ciphertexts, proofs and
    primary inputs We chose to sign the hash of the proof for modularity (to
    use the same code regardless of whether GROTH16 or PGHR13 proof system is
    chosen), and sign the hash of the ciphers and inputs for consistency.
    """
    # Parse the signature key pair
    vk = keypair.vk
    sk = keypair.sk

    # Format part of the public key as an hex
    y0_hex = hex_extend_32bytes("{0:0>64X}".format(int(vk[1][0])))
    y1_hex = hex_extend_32bytes("{0:0>64X}".format(int(vk[1][1])))

    # Encode and hash the verifying key and input hashes
    data_to_sign = encode_abi(
        ["bytes32", "bytes32", "bytes32"],
        [
            bytes.fromhex(y0_hex),
            bytes.fromhex(y1_hex),
            bytes.fromhex(hash_signature)
        ]
    )
    data_hex = sha256(data_to_sign).hexdigest()

    # Convert the hex digest into a field element
    h = int(data_hex, 16) % constants.ZETH_PRIME

    # Compute the signature sigma
    sigma = sk[1].n + h * sk[0].n % constants.ZETH_PRIME

    return sigma
```

We need to modify it to pass the message to be signed (the extra steps dealing with the vk should be done outside of this function) and only implement the signature algorithm inside this function @rrtoledo

Reply:
And as a side note the name `hash_signature` is misleading. It seems to represent the hash of the signature (ie: `hash(sigma)`) while it represents the hash of a subset of the `EthTx.data` (ie: ciphertexts, primary inputs, proof)

Reply:
To clarify this, we need to follow the key structure proposed in the paper of the signature used: https://eprint.iacr.org/2007/273.pdf page 15

> The KG algorithm runs pkg to get (ppk,psk),  runs skg(ppk,psk) to  get  (spk,ssk),  and  returns ppk‖spk as  the  public  key  and psk‖ssk as  the  secret  key.  The  remaining algorithms  simply  use  the  respective  algorithms  from  the  two-tier  signature  scheme.   That  is,  sign  algorithm SGN(psk‖ssk,M) returns S$←sgn(psk,ssk,M) [...]

Reply:
I find this notation a bit confusing because ppk is in psk.
 
I made this following change so far:
- `vk = (vk_1, vk_2)`
- `sk = (sk_1, sk_2, vk_1, vk_2)`
- Sign(sk, m)